### PR TITLE
Refactor buffer slice method

### PR DESCRIPTION
**Pull request by Achie72**
In iots_module_buffer.c the Slice method contained redundant code, which was moved out into it's seperate function now called `bufferslice_idx_normalizer`. The lenght was asked multiple times by function, so that was moved to the beginning, and saved in it's own variable.

IoT.js-DCO-1.0-Signed-off-by: Bela Toth tbela@inf.u-szeged.hu

### CODE REVIEW

File: src/modules/iotjs_module_buffer.c [hunk 1]
```diff
@@ -174,6 +174,22 @@ size_t iotjs_bufferwrap_copy(iotjs_bufferwrap_t* bufferwrap, const char* src,
   return iotjs_bufferwrap_copy_internal(bufferwrap, src, 0, len, 0);
 }
 
+static size_t bufferslice_idx_setter(int64_t point, size_t len) {
```
Comment (galpeter) [inline src/modules/iotjs_module_buffer.c]: Perhaps a better name would be `buffer_index_normalizer` and the parameters could be named `index` and `max_length`

File: src/modules/iotjs_module_buffer.c [hunk 2]
```diff
@@ -174,6 +174,22 @@ size_t iotjs_bufferwrap_copy(iotjs_bufferwrap_t* bufferwrap, const char* src,
...
+    if ((size_t)(-point) > len) {
+      idx = SIZE_MAX;
+    } else {
+      idx = (size_t)point + len;
+    }
+  } else {
+    idx = (size_t)point;
+  }
+
+
```
Comment (galpeter) [inline src/modules/iotjs_module_buffer.c]: There is no need for the double empty line here, one is enough.

File: src/modules/iotjs_module_buffer.c [hunk 3]
```diff
@@ -174,6 +174,22 @@ size_t iotjs_bufferwrap_copy(iotjs_bufferwrap_t* bufferwrap, const char* src,
...
+  idx = bound_range(idx, 0, len);
+  return idx;
```
Comment (galpeter) [inline src/modules/iotjs_module_buffer.c]: Just directly return with the bound_range value: `return bound_range(idx, 0, len);`

File: src/modules/iotjs_module_buffer.c [hunk 4]
```diff
@@ -348,31 +364,9 @@ JS_FUNCTION(Slice) {
...
-      end_idx = (size_t)end + len;
-    }
-  } else {
-    end_idx = (size_t)end;
-  }
-  end_idx = bound_range(end_idx, 0, iotjs_bufferwrap_length(buffer_wrap));
-
+  size_t len = iotjs_bufferwrap_length(buffer_wrap);
+  start_idx = bufferslice_idx_setter(start, len);
+  end_idx = bufferslice_idx_setter(end, len);
```
Comment (galpeter) [inline src/modules/iotjs_module_buffer.c]: You can add the variable definitions to these lines, no need to have them in a separate line, for example:
```c
size_t start_idx =...;
size_t end_idx = ...;
```

File: src/modules/iotjs_module_buffer.c [hunk 5]
```diff
@@ -174,6 +174,20 @@ size_t iotjs_bufferwrap_copy(iotjs_bufferwrap_t* bufferwrap, const char* src,
   return iotjs_bufferwrap_copy_internal(bufferwrap, src, 0, len, 0);
 }
 
+static size_t index_normalizer(int64_t index, size_t max_lenght) {
```
Comment (galpeter) [inline src/modules/iotjs_module_buffer.c]: There is one typo: `max_lenght` should be `max_length`

**Comment by yichoi** [comment]
LGTM

**Pull request merged**